### ISSUE TITLE
New rule: no-useless-empty-export

### ISSUE THREAD

**Issue by remcohaszing** [typescript-eslint/typescript-eslint#2757]
In TypeScript, modules are TypeScript files that contain an import or export statement. If a file is not a module, it will be ignored by TypeScript. Sometimes a file doesn’t contain imports or exports for legitimate reasons. In this case, an empty export must be added. I.e. to augment an interface:

```ts
export {};

declare module '@appsemble/sdk' {
  interface Parameters {
    interval: number;
  }

  interface EventEmitters {
    interval: never;
  }
}
```

As soon as the file is modified to contain actual imports or exports, the empty export can be removed.

I’d like to propose a rule that reports empty exports if the file contains other imports / exports. This can be autofixable by removing the empty export. I’m willing to implement this.

**Comment by remcohaszing** [comment]
Perhaps even an option could be added to make an export required if no imports are present. Maybe this would be even better.

```yaml
rules:
  @typescript-eslint/valid-module:
    - error
      # Removes an empty export if there are other imports or exports. Autofix removes the empty export.
    - allowUselessExport: false
      # Requires an empty export if there are no other imports or exports. Autofix adds an empty export to the end of the file.
      requireModuleSyntax: true
```

**Comment by bradzacher** [comment]
> If a file is not a module, it will be ignored by TypeScript

Could you explain what you mean by this?
This hasn't been my experience with TypeScript.

**Comment by remcohaszing** [comment]
From the TypeScript Handbook [Modules](https://www.typescriptlang.org/docs/handbook/modules.html) page:

> In TypeScript, just as in ECMAScript 2015, any file containing a top-level import or export is considered a module. Conversely, a file without any top-level import or export declarations is treated as a script whose contents are available in the global scope (and therefore to modules as well).

I don’t know the exact details of what this means. I do know a specific situation how it impacts my project.

Clone and setup [Appsemble](https://gitlab.com/appsemble/appsemble) and open it with a TypeScript editor:

```sh
git clone https://gitlab.com/appsemble/appsemble.git
cd appsemble
yarn
code .
```

Now open [blocks/timer/src/index.ts](https://gitlab.com/appsemble/appsemble/-/blob/master/blocks/timer/src/index.ts) and [blocks/timer/block.ts](https://gitlab.com/appsemble/appsemble/-/blob/master/blocks/timer/block.ts)

```sh
code blocks/timer/src/index.ts
code blocks/timer/block.ts
```

Everything should be fine. `block.ts` contains an `export {}` statement. If this is removed, the file is no longer considered to be a module by TypeScript. An error will appear in `src/index.ts`.

After reading [this article I just googled](https://malcolmkee.com/today-i-learnt/script-vs-module/) I believe scripts can’t declare modules.

Adding any import or export back to the file, will make TypeScript stop complaining. The only logical thing to do is add an empty export. However, this statement is useless as soon as another import or export is added.

**Comment by bradzacher** [comment]
I'm personally of the opinion that you should never `declare` in anything except for `.d.ts` files (aside from a few edge cases).

Doubly true for declaring modules.

`import`ing a random file to declare a module so that you can import another file seems like a really bad pattern?

OTOH if you declare a `.d.ts` file and put it in the right place, then it becomes an ambient module declaration file, and isn't required to be imported.

**Comment by remcohaszing** [comment]
Declaring a module is just an example I ran into. The actual issue is about TypeScript behaving differently for modules and scripts.

For example the following exports an interface:

```ts
export interface Foo {}
```

However, the following declares a global interface:

```ts
interface Foo {}
```

This ESLint rule would avoid the (probably accidental) global interface declaration by autofixing this to:

```ts
interface Foo {}

export {};
```

Now the developer using a consuming module will be reminded they forgot to export and import `Foo`. The developer should manually change it to the following:

```ts
export interface Foo {}

export {};
```

Now the ESLint rule should complain the `export {}` statement is useless and the autofix should remove it.

**Comment by IanVS** [comment]
Note, TypeScript 4.7 is going to allow more control over detection of modules vs scripts: https://devblogs.microsoft.com/typescript/announcing-typescript-4-7-beta/#control-over-module-detection
